---
title: "Genus_LPS_ISGs"
author: "Guannan Shen"
date: "May 29, 2019"
output: 
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: no
  word_document:
    toc: yes
    toc_depth: '5'
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/Omics_Integration/DataRaw/hiv_infected_un/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
```


```{r libs, echo = FALSE, warning= F}
######################
## Set up workspace
######################
rm(list = ls())
library(openxlsx)
library(tidyverse)
library(magrittr)
library(SmCCNet)
library(mixOmics)
library(r.jive)

options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
## not in function
'%nin%' <- Negate('%in%')
## left hand side vs right hand side
`%||%` <- function(lhs, rhs) {
  if (!is.null(lhs)) {  lhs } 
  else {  rhs }
}

# ######## clean memory ######################
# rm(list = ls())
# gc()
# is(dds)
# slotNames(dds)
```


# Omics Integration and Network Analysis

## Data
Clinical parameters: LPS (pg/ml).  
Genus level microbiome data, filtered via tidi_MIBI.  
RNAseq data of core-ISGs genes list, TMM normalized and regularized log transformed.  
26 subjects in common.  
252 genes (246 genes after filtering) and 19 OTUS (Including "Other") at the Genus level.  
The microbiome data was filtered with a stringent criteria: prevalence 30% and RA (relative abundance) 10%, with this cutoff, Remaining OTUS: 23  (Including "Other") at the Species level.  
Taxa which are present in a minority of the population were collapsed into a single “other” category.  
IFN-Beta specific genes contains 424 genes and 406 genes remain after filtering.  


```{r importdata}
# source to get the load_filtered_micro_level function
# this is to get centered log ratio transformed relative abundance with origincal dataset in default dir
source("~/Documents/gitlab/Omics_Integration/Code/5_29_Generate_filtered_Data_Microbiome.R")
# load filtered the microbiome data
# c("phylum", "order", "family", "genus", "species" )
gen_30_10 <- load_filtered_micro_level("genus", 30, 10, "Ubuntu")
# load_filtered_micro_level("species", 30, 10, "Ubuntu")

# gen_30_10 contains
# mibi_filter_ra, mibi_filter_clr, data_ra, lib_s (filtered ra, filtered clr, all ra, lib_size) 
# relative abundance has been multiplied by 100
micro_clr <- gen_30_10[[2]] %>% as.data.frame()
# rescale to mean 0 and variance 1 
mibi <- rescale_microbiome(micro_clr)

# genelist contains Gene_ID, Symbol
dir = "~/Documents/gitlab/Omics_Integration/DataRaw/hiv_infected_un/"
## import gene lists ##
isgs <- as.data.frame(read.delim( paste(dir, "coreISG", sep = "" )) )

genesbeta <- as.data.frame(read.delim( paste(dir, "genesbeta", sep = "" )) )

# clean and transform transcriptome data, subset of genes
source("~/Documents/gitlab/Omics_Integration/Code/6_5_clean_transcriptome.R")
# get gene symbols 
isgs_symbol = rescaled_rna(isgs, rlog = T)[[2]] %>% dplyr::select(Symbol)
# rlog transform or not
isgs_rlog <- rescaled_rna(isgs, rlog = T)[[1]]
# the name isgs covers the genelists isgs
isgs <- rescaled_rna(isgs, rlog = F)[[1]]

# check ID
sum(rownames(mibi) != rownames(isgs))

##### clinical data #########
source("~/Documents/gitlab/Omics_Integration/Code/6_5_clean_clinical.R")
clin <- rescaled_cli()
sum(clin$ID != rownames(isgs) )
print("All of the three dataset have the same order of samples")

# LPS is used here
# there is a missing value in LPS 
sum(is.na( clin$LPS ))
sum(is.na(clin$`IL-6`))
sum(is.na(clin$CD4_Counts ))

# Num 16 subjest is missing 
which(is.na( clin$LPS ))
n_na = which(is.na( clin$LPS ))
clin$ID[is.na( clin$LPS )]

#### reference plots ##################
source("~/Documents/gitlab/Omics_Integration/Code/ref_plots.R")

## compare variance with rlog or not
# Wrong! You can not make across genes comparison (within subject) by Transcriptome Data
# hist_var(isgs_rlog, "Within-subject Variance: Core ISGs rlog")
# hist_var(isgs, "Within-subject Variance: Core ISGs")

# # make data of variance and correlations 
# # data to compare within-subject variance
# data =  rbind( get_var(isgs, "No rlog"), 
#                get_var(isgs_rlog, "rlog") )
#                
# #
# box_values_group(data, "Core ISGs Preprocessed: Within-subject Variance")
# density_values_group(data, "Core ISGs Preprocessed: Within-subject Variance")

####### data to compare correlations ######
# no missing 
anyNA(mibi)
anyNA(isgs)
######################## need a function for this ###############################3
# use = "complete.obs" or get rid of subject 16
# tmp = stats::cor(mibi, isgs, use = "pairwise.complete.obs", method = "pearson)

df1 = get_corr(mibi, isgs, "Microbiome:Core ISGs")

clin_para = clin %>% select(CD4_Counts) %>% na.omit()
# stats::cor(clin$LPS, mibi, use = "pairwise.complete.obs", method = "pearson")
df2 = get_corr(clin_para, mibi, "CD4_Counts:Microbiome")
df3 = get_corr(clin_para, isgs, "CD4_Counts:Core ISGs")
data = rbind(df1, df2, df3)

box_values_group(data, "Pearson Correlations")
density_values_group(data, "Pearson Correlations")

clin_para = clin %>% select(`IL-6`) %>% na.omit()
# stats::cor(clin$LPS, mibi, use = "pairwise.complete.obs", method = "pearson")
df2 = get_corr(clin_para, mibi, "IL-6:Microbiome")
df3 = get_corr(clin_para, isgs, "IL-6:Core ISGs")
data = rbind(df1, df2, df3)

box_values_group(data, "Pearson Correlations")
density_values_group(data, "Pearson Correlations")

# the final clinical parameter to use
clin_para = clin %>% select(LPS) %>% na.omit()
df1 = get_corr(mibi[-n_na, ], isgs[-n_na, ], "Microbiome:Core ISGs")
# stats::cor(clin$LPS, mibi, use = "pairwise.complete.obs", method = "pearson")
df2 = get_corr(clin_para, mibi[-n_na, ], "LPS:Microbiome")
df3 = get_corr(clin_para, isgs[-n_na, ], "LPS:Core ISGs")
# plots
data = rbind(df1, df2, df3)

box_values_group(data, "Pearson Correlations")
density_values_group(data, "Pearson Correlations")
```

```{r smccnet}
#######  Unweighted SmCCA
# SubsamplingNum Number of feature subsamples. Default is 1000. Larger number leads to more
# accurate results, but at a higher cost.
# this sample size, 25, the paper use 4-fold 
sam_size = nrow(clin_para)
# check ID order again
sum(rownames(mibi) != rownames(isgs_rlog) )
sum(rownames(isgs_rlog) != clin$ID)
# where the missing is 
n_na = which(is.na( clin$LPS ))

#######  Unweighted SmCCA ###########
# SubsamplingNum Number of feature subsamples. Default is 1000. Larger number leads to more
# accurate results, but at a higher cost.
# this sample size, 25, the paper use 4-fold 
# n for p1 and all names
# isgs_symbol
n <-  ncol(isgs_rlog)
colnames(isgs_rlog) <- isgs_symbol$Symbol
names <- c(colnames(isgs_rlog), colnames(mibi))
# corre
x <- cbind(isgs_rlog, mibi)
corr <- cor(x, method = "pearson")

dim(mibi)

## il-6
il6 = clin %>% select(`IL-6`) %>% na.omit() %>% as.matrix() %>% as.vector()
set.seed(123)
W1 <- getRobustPseudoWeights(isgs_rlog, mibi, Trait = il6, Lambda1 = 0.01,
      Lambda2 = 0.01, s1 = 0.7, s2 = 0.9, NoTrait = FALSE, FilterByTrait = FALSE,
      SubsamplingNum = 2000, CCcoef = NULL, trace = FALSE)
dim(W1)
abar <- getAbar(W1, P1 = n, FeatureLabel = names)
modules <- getMultiOmicsModules(abar, P1 = n, CutHeight = 1 - 0.1^10, PlotTree = TRUE)
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = 1, P1 = n, FeatureLabel = names)

## cd4 counts

cd4 = clin %>% select(CD4_Counts) %>% na.omit() %>% as.matrix() %>% as.vector()
set.seed(123)
W1 <- getRobustPseudoWeights(isgs_rlog, mibi, Trait = cd4, Lambda1 = 0.01,
      Lambda2 = 0.01, s1 = 0.7, s2 = 0.9, NoTrait = FALSE, FilterByTrait = FALSE,
      SubsamplingNum = 2000, CCcoef = NULL, trace = FALSE)
dim(W1)
abar <- getAbar(W1, P1 = n, FeatureLabel = names)
modules <- getMultiOmicsModules(abar, P1 = n, CutHeight = 1 - 0.1^10, PlotTree = TRUE)
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = 1, P1 = n, FeatureLabel = names)

## lps
# lps %>% na.omit()
lps = clin %>% select(LPS)  %>% as.matrix() %>% as.vector()
length(lps)
set.seed(123)
W1 <- getRobustPseudoWeights(isgs_rlog, mibi, Trait = lps, Lambda1 = 0.01,
      Lambda2 = 0.01, s1 = 0.7, s2 = 0.9, NoTrait = FALSE, FilterByTrait = FALSE,
      SubsamplingNum = 2000, CCcoef = NULL, trace = FALSE)
dim(W1)
abar <- getAbar(W1, P1 = n, FeatureLabel = names)
modules <- getMultiOmicsModules(abar, P1 = n, CutHeight = 1 - 0.1^10, PlotTree = TRUE)
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = 1, P1 = n, FeatureLabel = names)

############## weighted SmCCA CCcoef = c(1, 5, 5) ##############

set.seed(123)
W1 <- getRobustPseudoWeights(isgs_rlog, mibi, Trait = il6, Lambda1 = 0.01,
      Lambda2 = 0.01, s1 = 0.7, s2 = 0.9, NoTrait = FALSE, FilterByTrait = FALSE,
      SubsamplingNum = 2000, CCcoef = c(1, 10, 2), trace = FALSE)
dim(W1)
abar <- getAbar(W1, P1 = n, FeatureLabel = names)
modules <- getMultiOmicsModules(abar, P1 = n, CutHeight = 1 - 0.1^10, PlotTree = TRUE)
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = 1, P1 = n, FeatureLabel = names)

# change the cutheight
modules <- getMultiOmicsModules(abar, P1 = n, CutHeight = 0.5, PlotTree = TRUE)
modules
# ModuleIdx Index for the module to be plotted. It can not exceed the length of multiOmicsModule
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = 1, P1 = n, FeatureLabel = names)


```
