---
title: "Gene Lists and LPS"
author: "Guannan Shen"
date: "June 11, 2019"
output: 
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: no
  word_document:
    toc: yes
    toc_depth: '5'
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/Omics_Integration/DataRaw/hiv_infected_un/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
```


```{r libs, echo = FALSE, warning= F}
######################
## Set up workspace
######################
rm(list = ls())
library(openxlsx)
library(tidyverse)
library(magrittr)

library(caret)   # for cross validation

library(SmCCNet)

options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
# not in function
'%nin%' <- Negate('%in%')
# is nothing
isnothing = function(x) {
  # test against NULL NA NaN
  any(is.null(x))  || any(is.na(x))  || any(is.nan(x)) 
}

## left hand side vs right hand side
'%||%' <- function(lhs, rhs) {
  if ( !isnothing(lhs) ) {  lhs } 
  else if ( !isnothing(rhs) ){  rhs }
  else { print("Both sides contain NULL, NA or NaN!")}
}

# ######## clean memory ######################
# rm(list = ls())
# gc()
# is(dds)
# slotNames(dds)
```

# Data and Functions

```{r importdata}
5 %||% NULL
```