---
title: "Gene Lists and LPS"
author: "Guannan Shen"
date: "June 11, 2019"
output: 
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: no
  word_document:
    toc: yes
    toc_depth: '5'
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/Omics_Integration/DataRaw/hiv_infected_un/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
```


```{r libs, echo = FALSE, warning= F}
######################
## Set up workspace
######################
rm(list = ls())
library(openxlsx)
library(tidyverse)
library(magrittr)
library(BiocStyle)
library(caret)   # for cross validation

library(SmCCNet)

options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()

# directory, Ubuntu 
dir = "~/Documents/gitlab/Omics_Integration/"
# small functions, %nin%, %||%, isnothing
source( paste0(dir, "Code/small_fun.R") )

# ######## clean memory ######################
# rm(list = ls())
# gc()
# is(dds)
# slotNames(dds)
```

# Data and Functions

```{r importdata}
# source functions 
# source to get the load_filtered_micro_level function to get clr of RA
source( paste0(dir, "Code/5_29_Generate_filtered_Data_Microbiome.R") )
# clean and transform transcriptome data, subset of genes
source( paste0(dir, "Code/6_5_clean_transcriptome.R") )
# clinical data 
source( paste0(dir, "Code/6_5_clean_clinical.R") )
# diagnostic plots and tables
source( paste0(dir, "Code/ref_plots.R") )

########## Datasets #############
#### phenotype contains ID
clin <- rescaled_cli() 
# import gene lists #
isgs <- as.data.frame(read.delim( paste0(dir, "DataRaw/hiv_infected_un/coreISG")) )
genesbeta <- as.data.frame(read.delim( paste0(dir, "DataRaw/hiv_infected_un/genesbeta" )) )
# LPS
# Num 16 subjest is missing 
which(is.na( clin$LPS ))

#### Transcriptome
# get gene symbols 
rna_isgs <- rescaled_rna(isgs, rlog = T)
isgs_symbol = rna_isgs[[2]] %>% dplyr::select(Symbol)
# rlog transform or not
isgs_rlog <- rna_isgs[[1]]

# ifn-beta
genesbeta <- as.data.frame(read.delim( paste0(dir, "DataRaw/hiv_infected_un/genesbeta" )) )
rna_genesbeta <- rescaled_rna(genesbeta, rlog = T)
# rlog transform or not
genesbeta_rlog <- rna_genesbeta[[1]]
# names
colnames(genesbeta_rlog) <- rna_genesbeta[[2]]$Symbol

#### Microbiome
gen_30_10 <- load_filtered_micro_level_samples("genus",  prevalence = 30, RA = 10, wd = "Ubuntu")
micro_clr <- gen_30_10[[2]] %>% as.data.frame()
# rescale to mean 0 and variance 1 
mibi <- rescale_microbiome(micro_clr)
#
anyNA(mibi)
anyNA(isgs_rlog)


```

# Density Plots: Hints for weighted SmCCNet
For LPS, the measurement of subject 16 is missing.  
Skewness:  
> Although SmCCNet does not require normality, it calculates the Pearson correlation between
linear combinations of omics features and the phenotype, which assumes finite variances and
finite covariance. It is necessary to include a transformation if the data are skewed.

```{r ref_weights}
# names
colnames(isgs_rlog) <- isgs_symbol$Symbol
# where na in LPS
n_na = which(is.na( clin$LPS ))
# get hints ideas
# the final clinical parameter to use
LPS= clin %>% select(LPS) %>% na.omit() 
df1 = get_corr(mibi[-n_na, ], isgs_rlog[-n_na, ], "Microbiome:Core ISGs")
# stats::cor(clin$LPS, mibi, use = "pairwise.complete.obs", method = "pearson")
df2 = get_corr(LPS, mibi[-n_na, ], "LPS:Microbiome")
df3 = get_corr(LPS, isgs_rlog[-n_na, ], "LPS:Core ISGs")
# plots
data = rbind(df1, df2, df3)

box_values_group(data, "Pearson Correlations")
density_values_group(data, "Pearson Correlations")

# starting points of weights
print("From the boxplot and the Density Plot: Weights 10, 5, 1 for X1X2, LPS:Microbiome, LPS:Core ISGs")

# check skewness
check_skew(micro_clr, "Density Plot of CLR of Microbiome RA")
check_skew(rescaled_rna(isgs, rlog = F)[[3]], "Density Plot of without rlog Transcriptome")
check_skew(rna_isgs[[3]], "Density Plot of rlog Transcriptome")

```

# SmCCNet Parameters Engineering
If EdgeCut = 0 (default), then the full module network will be created.

```{r smccnet_para}
########## parameters ################
# by convention, the first Omics is transcriptome and the 2nd is the microbiome 
# this sample size, 25, the paper use 4-fold 
# missing
n_na = which(is.na( clin$LPS ))
# check ID order again
sum(rownames(mibi) != rownames(isgs_rlog) )
sum(rownames(isgs_rlog) != clin$ID)

# unchanged parameters
p1 <- ncol(isgs_rlog)
p2 <- ncol(mibi)
n <- nrow(isgs_rlog[-n_na, ])
AbarLabel <- c(colnames(cbind(isgs_rlog, mibi)))

# # parameters 
# K <- 4 # Number of folds in K-fold CV.
# CCcoef <- NULL # Unweighted version of SmCCNet.
# # Microbiome is the 2nd one, thus s2 is larger
# s1 <- 0.7; s2 <- 0.9 # Feature sampling proportions.
# SubsamplingNum <- 1000 # Number of subsamples.

# 4 fold Cross Validation error
t12 <- read.csv(file = paste0("~/Documents/gitlab/Omics_Integration/DataProcessed/YCore_ISGs_Genus4foldCV/", 
                            "TotalPredictionError.csv"))

t12 %>% plyr::arrange(CC.Pred..Error) %>% kable(., 
                            caption = "aggregated pseudo canonical correlations and Total Prediction Error")





```

## Run SmCCNet
The trimmed module (edge cut = 0.1) is shown below. If a full module does not contain any edge that passes the cut threshold, a message “No edge passes threshold” will be produced. To see all complete module, set edgeCut = 0.  
The Lambda1 and Lambda2 are the most important parameters.  

```{r SmCCNet_results}
# correlation 
x <- cbind(isgs_rlog[-n_na, ], mibi[-n_na, ])
corr <- cor(x, method = "pearson")
# Make LPS a vector
LPS = clin %>% select(LPS) %>% na.omit() %>% as.matrix() %>% as.vector()

############# Unweighted ####################
set.seed(123)
W1 <- getRobustPseudoWeights(isgs_rlog[-n_na, ], mibi[-n_na, ], Trait = LPS, Lambda1 = 0.3,
      Lambda2 = 0.15, s1 = 0.7, s2 = 0.9, NoTrait = FALSE, FilterByTrait = FALSE,
      SubsamplingNum = 1000, CCcoef = NULL, trace = FALSE)
dim(W1)
# robust similarity matrix 
abar <- getAbar(W1, P1 = p1, FeatureLabel = AbarLabel)
modules <- getMultiOmicsModules(abar, P1 = p1, PlotTree = T, CutHeight = 1 - 0.1^10)
for(idx in 1:length(modules)){
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = idx, P1 = p1, FeatureLabel = AbarLabel,
                      EdgeCut = 0.1)
}


############## weighted ##################
set.seed(123)
W1 <- getRobustPseudoWeights(isgs_rlog[-n_na, ], mibi[-n_na, ], Trait = LPS, Lambda1 = 0.3,
      Lambda2 = 0.15, s1 = 0.7, s2 = 0.9, NoTrait = FALSE, FilterByTrait = FALSE,
      SubsamplingNum = 1000, CCcoef = c(10,5,1), trace = FALSE)
dim(W1)
abar <- getAbar(W1, P1 = p1, FeatureLabel = AbarLabel)
modules <- getMultiOmicsModules(abar, P1 = p1)
for(idx in 1:length(modules)){
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = idx, P1 = p1, FeatureLabel = AbarLabel, 
                      EdgeCut = 0.1)
}


```

### Tuning Results 
```{r tuning}
################ core ISGs ################
# correlation 
x <- cbind(isgs_rlog[-n_na, ], mibi[-n_na, ])
corr <- cor(x, method = "pearson")
p1 <- ncol(isgs_rlog)
p2 <- ncol(mibi)
n <- nrow(isgs_rlog[-n_na, ])
AbarLabel <- c(colnames(cbind(isgs_rlog, mibi)))
####
set.seed(123)
W1 <- getRobustPseudoWeights(isgs_rlog[-n_na, ], mibi[-n_na, ], Trait = LPS, Lambda1 = 0.35,
      Lambda2 = 0.1, s1 = 0.8, s2 = 0.9, NoTrait = FALSE, FilterByTrait = FALSE,
      SubsamplingNum = 1000, CCcoef = NULL, trace = FALSE)
dim(W1)
# robust similarity matrix 
abar <- getAbar(W1, P1 = p1, FeatureLabel = AbarLabel)
modules <- getMultiOmicsModules(abar, P1 = p1, PlotTree = T, CutHeight = 1 - 0.1^10)
for(idx in 1:length(modules)){
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = idx, P1 = p1, FeatureLabel = AbarLabel,
                      EdgeCut = 0)
}


### weighted ####
set.seed(123)
W1 <- getRobustPseudoWeights(isgs_rlog[-n_na, ], mibi[-n_na, ], Trait = LPS, Lambda1 = 0.3,
      Lambda2 = 0.05, s1 = 0.8, s2 = 0.9, NoTrait = FALSE, FilterByTrait = FALSE,
      SubsamplingNum = 1000, CCcoef = c(10,1,10), trace = FALSE)
dim(W1)
abar <- getAbar(W1, P1 = p1, FeatureLabel = AbarLabel)
modules <- getMultiOmicsModules(abar, P1 = p1)
for(idx in 1:length(modules)){
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = idx, P1 = p1, FeatureLabel = AbarLabel, 
                      EdgeCut = 0)
}

################## Beta ISGs ##############
# correlation 
x <- cbind(genesbeta_rlog[-n_na, ], mibi[-n_na, ])
corr <- cor(x, method = "pearson")
p1 <- ncol(genesbeta_rlog)
p2 <- ncol(mibi)
n <- nrow(genesbeta_rlog[-n_na, ])
AbarLabel <- c(colnames(cbind(genesbeta_rlog, mibi)))
#######
set.seed(123)
W1 <- getRobustPseudoWeights(genesbeta_rlog[-n_na, ], mibi[-n_na, ], Trait = LPS, Lambda1 = 0.2,
      Lambda2 = 0.05, s1 = 0.8, s2 = 0.9, NoTrait = FALSE, FilterByTrait = FALSE,
      SubsamplingNum = 1000, CCcoef = NULL, trace = FALSE)
dim(W1)
# robust similarity matrix 
abar <- getAbar(W1, P1 = p1, FeatureLabel = AbarLabel)
modules <- getMultiOmicsModules(abar, P1 = p1, PlotTree = T, CutHeight = 1 - 0.1^10)
for(idx in 1:length(modules)){
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = idx, P1 = p1, FeatureLabel = AbarLabel,
                      EdgeCut = 0)
}


#### weighted ###
set.seed(123)
W1 <- getRobustPseudoWeights(genesbeta_rlog[-n_na, ], mibi[-n_na, ], Trait = LPS, Lambda1 = 0.1,
      Lambda2 = 0.05, s1 = 0.8, s2 = 0.9, NoTrait = FALSE, FilterByTrait = FALSE,
      SubsamplingNum = 1000, CCcoef = c(10,1,10), trace = FALSE)
dim(W1)
abar <- getAbar(W1, P1 = p1, FeatureLabel = AbarLabel)
modules <- getMultiOmicsModules(abar, P1 = p1)
for(idx in 1:length(modules)){
plotMultiOmicsNetwork(abar, corr, modules, ModuleIdx = idx, P1 = p1, FeatureLabel = AbarLabel, 
                      EdgeCut = 0)
}

```
## Double Check the Results
This is to interpret the results.

```{r check_microbiome, include=F}
# #
# print("Why always you?")
# taxon_name <- colnames(mibi)[4]
# print(taxon_name)
# taxon <- gen_30_10[[3]][37, -c(1,3,4)] %>% t
# df1 <- taxon[-1, ] %>% as.data.frame() %>% rownames_to_column("ID")
# colnames(df1)[2] <- taxon[1, 1]
# df <- merge(df1, clin[, 1:2], by = "ID")
# df[,2] <- as.numeric(df[,2])
#  p = ggplot(df, aes(x = Group, y = `Bacteria/Bacteroidetes/Bacteroidia/Bacteroidales/Prevotellaceae`, 
#                     fill = Group )) + 
#     geom_boxplot(width=0.3) +
#    # grey scale plot
#     scale_fill_grey(start = 0.2, end = 0.8) +
#    # get the quantiles annotation
#     stat_summary(geom="text", fun.y=quantile,
#                  aes(label=sprintf("%1.1f", ..y..) ), color = "black",
#                  position=position_nudge(x=0.33), size = 4) +
#     theme_bw() +
#     labs(caption = "Relative Abundance by HIV Status") 
#  
#  print(p)
```